crates/e2e-test-utils/src/testsuite/actions.rs

```diff
@@ -0,0 +1,383 @@
+//! Actions that can be performed in tests.
+
+use crate::testsuite::{BoxFuture, Environment};
+use alloy_primitives::{BlockNumber, B256, U256};
+use alloy_rpc_types_engine::{ExecutionPayload, PayloadAttributes};
+use eyre::Result;
+use std::sync::Arc;
+
+/// An action that can be performed on an instance.
+///
+/// Actions can produce results that are stored for later assertions.
+pub trait Action<I>: Send + 'static {
+    /// Executes the action
+    fn execute<'a>(&'a mut self, env: &'a Environment<I>) -> BoxFuture<'a, Result<ActionResult>>;
+}
+
+/// Simplified action container for storage in tests
+#[allow(missing_debug_implementations)]
+pub struct ActionBox<I>(Box<dyn Action<I>>);
+
+impl<I: 'static> ActionBox<I> {
+    /// Constructor for [`ActionBox`].
+    pub fn new<A: Action<I>>(action: A) -> Self {
+        Self(Box::new(action))
+    }
+
+    /// Executes an [`ActionBox`] with the given [`Environment`] reference.
+    pub async fn execute(mut self, env: &Environment<I>) -> Result<ActionResult> {
+        self.0.execute(env).await
+    }
+}
+
+/// Possible results from actions
+#[derive(Debug)]
+pub enum ActionResult {
+    /// No result to store
+    None,
+    /// Store a payload with its identifier
+    Payload(String, Arc<ExecutionPayload>),
+    /// Store a block hash with its identifier
+    BlockHash(String, B256),
+    /// Store a transaction hash with its identifier
+    TransactionHash(String, B256),
+    /// Store a generic value with its identifier
+    Value(String, U256),
+    /// Store a boolean result with its identifier
+    Bool(String, bool),
+}
+
+/// Advance a single block with the given transactions.
+#[derive(Debug)]
+pub struct AdvanceBlock {
+    /// The node index to advance
+    pub node_idx: usize,
+    /// Transactions to include in the block
+    pub transactions: Vec<Vec<u8>>,
+    /// Identifier for storing the result
+    pub result_id: String,
+}
+
+impl<I> Action<I> for AdvanceBlock
+where
+    I: Send + Sync + 'static,
+{
+    fn execute<'a>(&'a mut self, _env: &Environment<I>) -> BoxFuture<'a, Result<ActionResult>> {
+        Box::pin(async move {
+            // 1. Create a new payload with the given transactions
+            // 2. Execute forkchoiceUpdated with the new payload
+            // 3. Return the payload ID and block hash
+
+            Ok(ActionResult::None)
+        })
```

Comment:
I assume the way this should is is that the environment provides access to the node handle or some helper trait that gives us access to the rpc endpoint